File: question1/javascript/zigzag_linked_list.js
class Node {
  constructor(data = null) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor(data) {
    if (Array.isArray(data)) {
      data.forEach(value => this.add_node(value));
    } else {
      this.add_node(data);
    }
  }

  add_node(data) {
    const node = new Node(data);

    if (!this.head) {
      this.head = node;
    } else {
      let current = this.head;
      while (current.next !== null) {
        current = current.next;
      }
      current.next = node;
    }
  }

  print_list() {
    process.stdout.write('head -> ');
    let current = this.head;
    while (current != null) {
      process.stdout.write(current.data + ' -> ');
      current = current.next;
    }
    console.log('*null');
  }
  // start with a flag of true
  // if flag is true then expected current.next.data > current.data
  // else if flag false, we expect current.next.data < current.data
  // otherwise swap
  zigzag(flag = true) {
    let current = this.head;

    while (current.next !== null) {
      if (flag == true && current.next.data < current.data) {
        [current.next.data, current.data] = [current.data, current.next.data];
      } else if (flag == false && current.next.data > current.data) {
        [current.next.data, current.data] = [current.data, current.next.data];
      }
      flag = !flag;
      current = current.next;
    }
    this.print_list(); // after zig zag
  }
}

const LL = new LinkedList([1, 2, 3, 4, 5, 6, 7, 8, 9]);
LL.print_list();
LL.zigzag(false);

File: question1/python/zigzag_linked_list.py
class Node:
    data = None
    next = None

    def __init__(self, data):
        self.data = data


class LinkedList:
    head = None

    def __init__(self, data):
        if isinstance(data, list):
            for value in data:
                self.add_node(value)
        else:
            self.add_node(data)

    def add_node(self, value):
        node = Node(value)

        if self.head is None:
            self.head = node

        else:
            current = self.head
            while current.next is not None:
                current = current.next

            current.next = node

    def print_list(self):
        print("head -> ", end="")
        current = self.head
        while current is not None:
            print(str(current.data) + " -> ", end="")
            current = current.next
        print("*null")

    def zigzag(self, flag=True):
        # start with flag True
        # if flag is True then expect current.next.data > current.data
        # else flag is False expect current.next.data < current.data
        # otherwise swap
        current = self.head
        while current.next is not None:
            if flag and current.next.data < current.data:
                current.next.data, current.data = current.data, current.next.data
            elif not flag and current.next.data > current.data:
                current.next.data, current.data = current.data, current.next.data

            flag = not flag
            current = current.next

        self.print_list()


LL = LinkedList([1, 2, 3, 4, 5, 6, 7, 8, 9])

LL.print_list()
LL.zigzag(False)
